### ISSUE TITLE
Feature request: SQLite plugin/module

### ISSUE THREAD

**Issue by alexrainman** [NativeScript/NativeScript#924]
I know there's a third party plugin but i will love you guys to release one or validate the third party one as ready for production.

https://github.com/nathanaela/nativescript-sqlite

**Comment by dtopuzov** [comment]
We plan Verified Plugins Marketplace for {N}, stay tuned :)

**Comment by NathanaelA** [comment]
@alexrainman - I wrote the third party sqlite plugin; and I'm using this for a couple of my apps (and I have a couple other people who I know that are using it too).   If you find a bug in it let me know and I'll be sure to get it fixed.   Right now I'm working on getting the iOS side switched to using a cocoapod to simplify installation.   ;-)

**Comment by alexrainman** [comment]
@NathanaelA - i will try your plugin and keep you posted. I am planning to use {N} in some personal projects. Thanks.

**Comment by dtopuzov** [comment]
@NathanaelA, please provide simple demo app with {N} 1.4+ and we will verify your plugin.

**Comment by alexrainman** [comment]
It seems the component store is about to launch. Awesome!
I better get mine ready.

**Comment by NathanaelA** [comment]
@dtopuzov -- I'll see what I can do.

**Comment by valentinstoychev** [comment]
@alexrainman - yes we are making a "silent" launch next week :).

**Comment by NathanaelA** [comment]
@dtopuzov - I have update my sqlite plugin so that it is now a seemless install for iOS; and I have pushed a new repo https://github.com/NathanaelA/nativescript-sqlite-example that has a simple example application using sqlite.

**Comment by alexrainman** [comment]
It works on Android like a charm but it doesn't in iOS. When i run the app i see this issue in the console:

```
2015-10-27 22:31:18.954 node[51933:3100750] [MT] DVTAssertions: Warning in /Library/Caches/com.apple.xbs/Sources/DVTFrameworks/DVTFrameworks-9071/DVTFoundation/PlugInArchitecture/PlugInManager/DVTPlugInManager.m:257
Details:  Requested but did not find extension point with identifier Xcode.DVTFoundation.DevicePlatformMapping
Object:   <DVTPlugInManager: 0x10230c740>
Method:   -extensionPointWithIdentifier:
Thread:   <NSThread: 0x1058047e0>{number = 1, name = main}
```

Doing some debugging, this part seems not to work in iOS:

```
if (!sqlite.exists(dbname)) {
   sqlite.copyDatabase(dynamo);
}
```

So, db is null.

**Comment by alexrainman** [comment]
Add button error in iOS:

```
file:///app/main-page.js:29:11: JS ERROR TypeError: null is not an object (evaluating 'db.execSQL')
1   0xc4610 NativeScript::FFICallback<NativeScript::ObjCMethodCallback>::ffiClosureCallback(ffi_cif*, void*, void**, void*)
2   0x6a0946 ffi_closure_SYSV_inner
3   0x6a0392 .LCFI7
4   0x293f7cd -[NSObject performSelector:withObject:withObject:]
5   0x13f3a40 -[UIApplication sendAction:to:from:forEvent:]
6   0x13f39d2 -[UIApplication sendAction:toTarget:fromSender:forEvent:]
7   0x153413a -[UIControl sendAction:to:forEvent:]
8   0x1534557 -[UIControl _sendActionsForEvents:withEvent:]
9   0x15337c1 -[UIControl touchesEnded:withEvent:]
10  0x144bcaa -[UIWindow _sendTouchesForEvent:]
11  0x144c786 -[UIWindow sendEvent:]
12  0x140a681 -[UIApplication sendEvent:]
13  0x141aab8 _UIApplicationHandleEventFromQueueEvent
14  0x13ee2e7 _UIApplicationHandleEventQueue
15  0x2bc206f __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__
16  0x2bb7b7d __CFRunLoopDoSources0
17  0x2bb70d8 __CFRunLoopRun
18  0x2bb6a5b CFRunLoopRunSpecific
19  0x2bb688b CFRunLoopRunInMode
20  0x48542c9 GSEventRunModal
21  0x4854106 GSEventRun
22  0x13f20b6 UIApplicationMain
23  0x6a02ad .LCFI1
24  0x6a088f ffi_call
25  0x93ae5 NativeScript::FFICall::call(JSC::ExecState*)
26  0x44795e JSC::LLInt::setUpCall(JSC::ExecState*, JSC::Instruction*, JSC::CodeSpecializationKind, JSC::JSValue, JSC::LLIntCallLinkInfo*)
27  0x443e46 llint_slow_path_call
28  0x44c643 llint_entry
29  0x44c653 llint_entry
30  0x44c653 llint_entry
31  0x44c653 lint_entry
```

**Comment by NathanaelA** [comment]
You have:
```js
if (!sqlite.exists(dbname)) {
   sqlite.copyDatabase(dynamo);
}
```
I believe the issue is that either dbname is null or dynamo is null.  They both should be the same variable.   The idea behind this code is to check to see if that database exists; and if it doesn't then to copy it.  So they variables need to be the same.  :grinning:

**Comment by alexrainman** [comment]
I know what is this code for. I do the same thing in my Xamarin apps. I like to ship a pre-made sqlite db.

This is your code running here my friend. It works on android but not ios :)

**Comment by alexrainman** [comment]
That dynamo thing is maybe copy and paste stuff :)

**Comment by NathanaelA** [comment]
What version of ios NativeScript runtimes are you using?

**Comment by alexrainman** [comment]
1.4.0

**Comment by NathanaelA** [comment]
Hmm, I can't duplicate your issue (but I did find another ios sqlite issue that I just fixed).  Can you try downloading the nativescript-sqlite-example app, install the requirements and then test it.  It installs a database and I just tested it on my mac.

**Comment by alexrainman** [comment]
Same issue my brother. Works on Android but not in iOS. Same errors as before :(

**Comment by teobugslayer** [comment]
@alexrainman,

 {N} CLI dev here. In our [1.4.3 release](https://github.com/NativeScript/nativescript-cli/blob/master/CHANGELOG.md) we fixed plenty of iOS related issues. I strongly advise to update.

**Comment by alexrainman** [comment]
I have CLI 1.4.3

**Comment by alexrainman** [comment]
OS X El Capitan, XCode 7.1, tns ios 1.4.x, tns cli 1.4.3

**Comment by NathanaelA** [comment]
I just did this on my Mac 10.9 box and it worked perfectly.
Reset the simulator, so it was totally clean then
Steps:
```
git clone http://github.com/NathanaelA/nativescript-sqlite-example
cd nativescript-sqlite-example
npm install tns-core-modules
tns platform add ios
tns plugin add nativescript-sqlite
tns run ios --emulator
```
Running: 
  tns version 1.43
  tns-core: 1.40
  tns-ios: 1.41
  nativescript-sqlite: 1.0.2

It started up as expected with my fully populated database, and let me add new names.

**Comment by alexrainman** [comment]
I will try this and let you know. Thanks

**Comment by alexrainman** [comment]
Same problem.

```
2015-10-29 17:59:33.706 node[55446:3625841] [MT] DVTAssertions: Warning in /Library/Caches/com.apple.xbs/Sources/DVTFrameworks/DVTFrameworks-9071/DVTFoundation/PlugInArchitecture/PlugInManager/DVTPlugInManager.m:257
Details:  Requested but did not find extension point with identifier Xcode.DVTFoundation.DevicePlatformMapping
Object:   <DVTPlugInManager: 0x103936ff0>
Method:   -extensionPointWithIdentifier:
Thread:   <NSThread: 0x10390d350>{number = 1, name = main}
```

**Comment by alexrainman** [comment]
Some forum talking about the same thing but with ionic http://forum.ionicframework.com/t/warning-when-emulating-ios/33822

**Comment by NathanaelA** [comment]
I have a theory.  I need to you go to the YOUR_APP/node_modules/nativescript-sqlite/platforms/ios
folder.  In that folder is a module.modulemap open it up in a text editor and:
1. Try eliminating the **link "sqlite3"**
2. tns run ios --emulator
3. If that didn't work, put back the **link "sqlite3"** and remove the **link "libsqlite3"**
4. tns run ios --emulator

See if one of those is causing the issue.   This doesn't cause an issue on 10.9, but maybe it does on El Capitan.

**Comment by alexrainman** [comment]
None of them works :(

**Comment by alexrainman** [comment]
@NathanaelA something you may be thinking to add to your plugin is encryption. I think field based is the easier way to do it.

**Comment by NathanaelA** [comment]
@alexrainman - Ok, I cannot duplicate your issue.  I spent the last week downloading the new El Capitan, and then installed whatever the latest Xcode version is.  (My internet connection is that SLOW).  I have totally fresh machine with the fresh copy of El Capitan, installed Xcode from the store, installed Xcode command line tools, installed brew, installed latest version of node from nodejs, installed Cocoapods, Installed NativeScript (v1.5 as of this moment).

I used a totally clean copy of my SQLite example; did a **tns platform install ios** and then a **tns plugin add nativescript-sqlite** (so everything was freshly downloaded from the internet) -- then finally a **tns run ios --emulator** and it started up and the app worked fine.    

The plugin worked fine on my Maveriks and now my El Capitan box.    Something must be messed up in your environment; literally this machine was just done being built, and I built and tried the nativescript sample, then my sqlite sample and everything worked as expected.
